src/_ert/forward_model_runner/client.py

```diff
@@ -26,107 +22,139 @@ class ClientConnectionClosedOK(Exception):
+                    self._ack_event.set()
+                else:
+                    await self.process_message(raw_msg.decode("utf-8"))
+            except zmq.ZMQError as exc:
+                logger.debug(
+                    f"{self.dealer_id} connection to evaluator went down, reconnecting: {exc}"
+                )
+                await asyncio.sleep(1)
+                self.socket.connect(self.url)
 
-    async def get_websocket(self) -> ClientConnection:
-        return await connect(
-            self.url,
-            ssl=self._ssl_context,
-            additional_headers=self._additional_headers,
-            open_timeout=self.CONNECTION_TIMEOUT,
-            ping_timeout=self.CONNECTION_TIMEOUT,
-            ping_interval=self.CONNECTION_TIMEOUT,
-            close_timeout=self.CONNECTION_TIMEOUT,
-        )
+    async def _send(self, message: str, retries: int | None = None) -> None:
+        self._ack_event.clear()
 
-    async def _send(self, msg: AnyStr) -> None:
-        for retry in range(self._max_retries + 1):
+        backoff = 1
+        retries = retries or self.DEFAULT_MAX_RETRIES
+        while retries >= 0:
             try:
-                if self.websocket is None:
-                    self.websocket = await self.get_websocket()
-                await self.websocket.send(msg)
-                return
-            except ConnectionClosedOK as exception:
-                error_msg = (
-                    f"Connection closed received from the server {self.url}! "
-                    f" Exception from {type(exception)}: {exception!s}"
-                )
-                raise ClientConnectionClosedOK(error_msg) from exception
-            except (TimeoutError, InvalidHandshake, InvalidURI, OSError) as exception:
-                if retry == self._max_retries:
-                    error_msg = (
-                        f"Not able to establish the "
-                        f"websocket connection {self.url}! Max retries reached!"
-                        " Check for firewall issues."
-                        f" Exception from {type(exception)}: {exception!s}"
+                await self.socket.send_multipart([b"", message.encode("utf-8")])
+                try:
+                    await asyncio.wait_for(
+                        self._ack_event.wait(), timeout=self._ack_timeout
                     )
-                    raise ClientConnectionError(error_msg) from exception
-            except ConnectionClosedError as exception:
-                if retry == self._max_retries:
-                    error_msg = (
-                        f"Not been able to send the event"
-                        f" to {self.url}! Max retries reached!"
-                        f" Exception from {type(exception)}: {exception!s}"
+                    return
+                except TimeoutError:
+                    logger.warning(
+                        f"{self.dealer_id} failed to get acknowledgment on the {message}. Resending."
```

Comment:
What will happen if the event shows up after it times out? Will this cause it to hang, until maxretries is reached?
msg0 goes out, no one replies
msg1 goes out, ack0 returns
we return.             ack1 returns

Reply:
No wait. There is nothing to be left hanging. Getting an extra ACK after everything finished is handled. Very nice!